expired_tree_test.go

```diff
@@ -0,0 +1,243 @@
+package verkle
+
+import (
+	"bytes"
+	"errors"
+	"testing"
+)
+
+func TestInsertSameLeafNoExpired(t *testing.T) {
+	t.Parallel()
+
+	root := New()
+	if err := root.Insert(zeroKeyTest, testValue, 0, nil); err != nil {
+		t.Fatalf("error inserting: %v", err)
+	}
+
+	if err := root.Insert(oneKeyTest, testValue, 1, nil); err != nil {
+		t.Fatalf("error inserting: %v", err)
+	}
+
+	leaf, ok := root.(*InternalNode).children[0].(*LeafNode)
+	if !ok {
+		t.Fatalf("expected leaf node, got %T", root.(*InternalNode).children[0])
+	}
+
+	if !bytes.Equal(leaf.values[zeroKeyTest[StemSize]], testValue) {
+		t.Fatalf("expected value %x, got %x", testValue, leaf.values[zeroKeyTest[StemSize]])
+	}
+
+	if !bytes.Equal(leaf.values[oneKeyTest[StemSize]], testValue) {
+		t.Fatalf("expected value %x, got %x", testValue, leaf.values[oneKeyTest[StemSize]])
+	}
+
+	if leaf.lastPeriod != 1 {
+		t.Fatalf("expected last accessed to be 1, got %d", leaf.lastPeriod)
+	}
+}
+
+func TestInsertSameLeafExpired(t *testing.T) {
+	t.Parallel()
+
+	root := New()
+	if err := root.Insert(zeroKeyTest, testValue, 0, nil); err != nil {
+		t.Fatalf("error inserting: %v", err)
+	}
+
+	err := root.Insert(oneKeyTest, testValue, 2, nil)
+	if !errors.Is(err, errExpired) {
+		t.Fatalf("expected period expired error when inserting, got %v", err)
+	}
+
+	leaf, ok := root.(*InternalNode).children[0].(*LeafNode)
+	if !ok {
+		t.Fatalf("expected leaf node, got %T", root.(*InternalNode).children[0])
+	}
+
+	if !bytes.Equal(leaf.values[zeroKeyTest[StemSize]], testValue) {
+		t.Fatalf("expected value %x, got %x", testValue, leaf.values[zeroKeyTest[StemSize]])
+	}
+
+	if leaf.lastPeriod != 0 {
+		t.Fatalf("expected last accessed to be 0, got %d", leaf.lastPeriod)
+	}
+}
+
+func TestInsertDiffLeaf(t *testing.T) {
+	t.Parallel()
+
+	root := New()
+	if err := root.Insert(zeroKeyTest, testValue, 0, nil); err != nil {
+		t.Fatalf("error inserting: %v", err)
+	}
+
+	if err := root.Insert(ffx32KeyTest, testValue, 2, nil); err != nil {
+		t.Fatalf("error inserting: %v", err)
+	}
+
+	leaf0, ok := root.(*InternalNode).children[0].(*LeafNode)
+	if !ok {
+		t.Fatalf("expected leaf node, got %T", root.(*InternalNode).children[0])
+	}
+
+	leaff, ok := root.(*InternalNode).children[255].(*LeafNode)
+	if !ok {
+		t.Fatalf("expected leaf node, got %T", root.(*InternalNode).children[255])
+	}
+
+	if !bytes.Equal(leaf0.values[zeroKeyTest[StemSize]], testValue) {
+		t.Fatalf("expected value %x, got %x", testValue, leaf0.values[zeroKeyTest[StemSize]])
+	}
+
+	if !bytes.Equal(leaff.values[ffx32KeyTest[StemSize]], testValue) {
+		t.Fatalf("expected value %x, got %x", testValue, leaff.values[ffx32KeyTest[StemSize]])
+	}
+
+	if leaf0.lastPeriod != 0 {
+		t.Fatalf("expected last accessed to be 0, got %d", leaf0.lastPeriod)
+	}
+
+	if leaff.lastPeriod != 2 {
+		t.Fatalf("expected last accessed to be 2, got %d", leaff.lastPeriod)
+	}
```

Comment:
you should also try to insert a sibling into an expired leaf, I might have missed it, but it seems to me that this case isn't being tested.

e.g. insert(00...0000, 0), and then insert(00..1100, 2) and see if intermediate branch nodes are inserted.